numpy/lib/shape_base.py

```diff
@@ -766,7 +766,11 @@ def array_split(ary, indices_or_sections, axis=0):
     try:
         Ntotal = ary.shape[axis]
     except AttributeError:
-        Ntotal = len(ary)
+        #if ary is Array_like instead of an ndarray.
+        if (axis == 0):
+            Ntotal = len(ary)
+        else:
+            Ntotal = len(ary[0])
```

Comment:
Sorry for not reviewing earlier.  If you look closely at this fix, you will see that it relies on the input being a nested sequence (or nested list) which has exactly 2 dimensions.  So it will only work for `axis=1` (or -1 with 2D input).

The tricky thing about this code, is that it works even for inputs like pandas dataframes right now (even if they were not only 2D in this sense).  That is, objects which are not NumPy arrays at all and for which it is not OK to call `np.asanyarray()`.
So we cannot just convert to a NumPy array, without potentially braking things (could be fine, but would require a lot of additional thought).

If you look closely at the code, the important part is the call to:
```
sary = _nx.swapaxes(ary, axis, 0)
```
which will ensure that the input is a NumPy array if it was not one before.  My feeling would be to move that call to the very top.  After that we can just always use `len()` (or `sary.shape[0]`) and remove the try/except (it seems fine to assume that an object which defines `obj.swapaxes` will also define `shape` and `len` compatibily).

For tests, maybe a deeper nested sequence would be good additionally.  In principle it would be nice to test a dummy custom object that just defines `swapaxes`, `__getitem__` and `shape`/`__len__`.